src/data/transform.ts

```diff
@@ -0,0 +1,555 @@
+        ExcalidrawImageElement | ExcalidrawFrameElement
+      >;
+      if (!existingElement) {
+        console.error(`No element for end binding with id ${end.id} found`);
+      }
+    }
+    const endX =
+      end.x || excliadrawLinearElement.x + excliadrawLinearElement.width;
+    const endY = end.y || excliadrawLinearElement.y - height / 2;
+    const endType = existingElement ? existingElement.type : end.type;
+
+    if (endType) {
+      if (endType === "text") {
+        let text = "";
+        if (existingElement && existingElement.type === "text") {
+          text = existingElement.text;
+        } else if (end.type === "text") {
+          text = end.text;
+        }
+
+        if (!text) {
+          console.error(
+            `No text found for end binding text element for ${excliadrawLinearElement.id}`,
+          );
+        }
+        endBoundElement = newTextElement({
+          x: endX,
+          y: endY,
+          type: "text",
+          ...existingElement,
+          ...end,
+          text,
+        });
+        // to position the text correctly when coordinates not provided
+        mutateElement(endBoundElement, {
+          y: end.y || excliadrawLinearElement.y - endBoundElement.height / 2,
+        });
+      } else {
+        endBoundElement = newElement({
+          x: endX,
+          y: endY,
+          width,
+          height,
+          ...existingElement,
+          ...end,
+          type: endType,
+        }) as ExcalidrawBindableElement;
+      }
+
+      bindLinearElement(
+        excliadrawLinearElement,
+        endBoundElement as ExcalidrawBindableElement,
+        "end",
+      );
+    }
+  }
+  return {
+    linearElement: excliadrawLinearElement,
+    startBoundElement,
+    endBoundElement,
+  };
+};
+
+const excalidrawElements = (() => {
+  const res: ExcalidrawElement[] = [];
+  const elementMap = new Map<string, number>();
+
+  const add = (ele?: ExcalidrawElement) => {
+    if (!ele) {
+      return;
+    }
+    const index = elementMap.get(ele.id);
+    if (index !== undefined && index >= 0) {
+      res[index] = ele;
+    } else {
+      res.push(ele);
+      const index = res.length - 1;
+      elementMap.set(ele.id, index);
+    }
+  };
+  const clear = () => {
+    res.length = 0;
+    elementMap.clear();
+  };
+  const get = () => {
+    return res;
+  };
+  const hasElementWithId = (id: string) => {
+    const index = elementMap.get(id);
+    return index !== undefined && index >= 0;
+  };
+  return {
+    add,
+    clear,
+    get,
+    hasElementWithId,
+  };
+})();
+
+export const convertToExcalidrawElements = (
+  elements: ExcalidrawProgrammaticAPI["elements"],
+): ExcalidrawElement[] => {
+  excalidrawElements.clear();
+  if (!elements) {
+    return [];
+  }
+  elements.forEach((element) => {
+    let elementId = element.id || regenerateId(null);
+
+    // To make sure every element has a unique id
+    while (excalidrawElements.hasElementWithId(elementId)) {
```

Comment:
no not `nanoid` but `regenerateId` could since it appends `_copy` to original id which might already exist so this is to make sure we don't use same ids for different elements